rest: Ensure response body is fully read and closed before retry

MadhavJivrajani opened the pull request:
This commit refactors the retry logic to include resetting the
request body. The reset logic will be called iff it is not the
first attempt. This refactor is nescessary mainly because now
as per the retry logic, we always ensure that the request body
is reset *after* the response body is *fully* read and closed
in order to reuse the same TCP connection.
    
Previously, the reset of the request body and the call to read
and close the response body were not in the right order, which
leads to race conditions.


xref https://github.com/kubernetes/kubernetes/issues/108906 

Fixes a bug in our client retry logic that was not draining/closing the body prior to trying to reset the request body for a retry. According to go upstream, this is required to avoid races.

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
client-go: if resetting the body fails before a retry, an error is now surfaced to the user.
```

Continuing https://github.com/kubernetes/kubernetes/pull/109028

/priority critical-urgent
/assign @aojea @liggitt @smarterclayton 
/cc @dims @tkashem 
would something like this work?

MadhavJivrajani commented:
/sig api-machinery
/kind cleanup

MadhavJivrajani commented:
/skip

liggitt commented on staging/src/k8s.io/client-go/rest/with_retry.go:
```diff
@@ -372,6 +379,63 @@ func readAndCloseResponseBody(resp *http.Response) {
+	return r.err
+}
+
+type buffer struct {
```
can use `io.NopCloser(&bytes.Buffer{})` to do this

MadhavJivrajani replied:
yeah i had done this initially, but we need a `Write()` method to satisfy `io.Writer` for use in `io.TeeReader`

liggitt replied:
let me know when this is pushed, dropping the `type buffer struct`

MadhavJivrajani replied:
done @liggitt 
thanks

liggitt commented on staging/src/k8s.io/client-go/rest/with_retry.go:
```diff
@@ -372,6 +379,63 @@ func readAndCloseResponseBody(resp *http.Response) {
+type ResponseCloser struct {
+	once     sync.Once
+	resp     *http.Response
+	respCopy io.ReadWriteCloser
```
rather than storing respCopy as a member variable and counting on callers to remember to use it instead of resp.Body, what if we replace resp.Body inside DrainAndClose()?

liggitt commented on staging/src/k8s.io/client-go/rest/with_retry.go:
```diff
@@ -372,6 +379,63 @@ func readAndCloseResponseBody(resp *http.Response) {
+func NewResponseCloser(resp *http.Response) *ResponseCloser {
+	return &ResponseCloser{
+		resp:     resp,
+		respCopy: &buffer{b: &bytes.Buffer{}},
```
could defer allocation of this to when we actually need it in the `if r.resp.ContentLength <= maxBodySlurpSize` block

liggitt commented on staging/src/k8s.io/client-go/rest/with_retry.go:
```diff
@@ -372,6 +379,63 @@ func readAndCloseResponseBody(resp *http.Response) {
+		if r.resp.ContentLength <= maxBodySlurpSize {
+			var reader io.Reader = &io.LimitedReader{R: r.resp.Body, N: maxBodySlurpSize}
+			reader = io.TeeReader(reader, r.respCopy)
+			io.Copy(ioutil.Discard, reader)
```
could replace resp.Body with respCopy here

liggitt commented on staging/src/k8s.io/client-go/rest/request.go:
```diff
@@ -631,17 +631,25 @@ func (r *Request) Watch(ctx context.Context) (watch.Interface, error) {
+			closer := NewResponseCloser(resp)
+			defer closer.DrainAndClose()
+
+			// get the result out of the response if provided *before* IsNextRetry,
+			// since that has to drain/close the response before resetting the request body
+			var result Result
+			if resp != nil {
+				result = r.transformResponse(resp, req)
+			}
```
I like the idea of buffering the body into a copy in IsNextRetry... if we do that, maybe we can keep the transformResponse call in the original location?

liggitt commented on staging/src/k8s.io/client-go/rest/request.go:
```diff
@@ -749,12 +757,21 @@ func (r *Request) Stream(ctx context.Context) (io.ReadCloser, error) {
+				closer := NewResponseCloser(resp)
+				defer closer.DrainAndClose()
+
+				// get the result out of the response if provided *before* IsNextRetry,
+				// since that has to drain/close the response before resetting the request body
+				var result Result
+				if resp != nil {
+					result = r.transformResponse(resp, req)
+				}
```
same comment about keeping transformResponse in the original location if we buffer in IsNextRetry

liggitt commented on staging/src/k8s.io/client-go/rest/request.go:
```diff
@@ -877,8 +894,11 @@ func (r *Request) request(ctx context.Context, fn func(*http.Request, *http.Resp
+			closer := NewResponseCloser(resp)
+			defer closer.DrainAndClose()
 
+			respCopy := shallowCloneHTTPResponse(resp)
```
if we replace resp.Body internally in DrainAndClose(), I don't think we need to do this

liggitt commented on staging/src/k8s.io/client-go/rest/request.go:
```diff
@@ -887,11 +907,12 @@ func (r *Request) request(ctx context.Context, fn func(*http.Request, *http.Resp
 			}
 
-			f(req, resp)
+			respCopy.Body = closer.respCopy
```
if we replace resp.Body internally in DrainAndClose(), I don't think we need to do this

liggitt commented:
I like the bones of this approach. It only triggers when:

* we have a 5xx or 429 response or a retriable transport error
* the request has a non-nil, seekable body

The only failure mode I see is when both of the following are true:
1. The response body has more than 1kb of content (we only buffer if the body is < 1 kb)
2. The request.body.Seek() call fails

In that case, we return that we should *not* retry, the response handling function is called, and it fails to read the body because we already drained/closed it

Adding a unit test to exercise that exact scenario (with a dummy request body that fails Seek()) would be useful. I expect what would happen is the response handling function would not have response body content and would handle the response generically based on the status code. That seems reasonable to me.

liggitt commented on staging/src/k8s.io/client-go/rest/with_retry.go:
```diff
@@ -372,6 +379,63 @@ func readAndCloseResponseBody(resp *http.Response) {
+	}
+}
+
+func (r *ResponseCloser) DrainAndClose() error {
```
since we only need to do this buffering when called from inside prepareNextRetry, can take a param to make the buffering optional, and pass `false` when we call this everywhere except prepareNextRetry

aojea commented:
if our problem is to read the response body so we can rewind the request body for a retry, why don't just clone always the request body, instead of rewinding it we use a new one ... I think that will be much simpler and we don't have to deal with all these additional changes

liggitt commented:
> why don't just clone always the request body

because that doubles the cost of the response and can be very large in some scenarios. In 429/500 scenarios, we expect the response body to be very small

aojea commented:
> > why don't just clone always the request body
> 
> because that doubles the cost of the response and can be very large in some scenarios. In 429/500 scenarios, we expect the response body to be very small

let me try to recap, because I'm not sure I follow you.

There are two "body"s, response and request, we have to deal here with the problem of "rewinding" the request body when we have to retry, but ...

1. we retry only on network errors, that means there is never a response.
2. We also only retry only on GETs ,if I understand the code correctly, that means the request body is never big, we can copy and not have to rewind

liggitt commented:
> > > why don't just clone always the request body
> > 
> > 
> > because that doubles the cost of the response and can be very large in some scenarios. In 429/500 scenarios, we expect the response body to be very small
> 
> let me try to recap, because I'm not sure I follow you.

oops, I misread, you suggested always cloning the *request* body, not the response body

> we retry only on network errors, that means there is never a response.

I'm pretty sure that's not true, we can retry in response to a 429 from the server

> We also only retry only on GETs ,if I understand the code correctly, that means the request body is never big, we can copy and not have to rewind

I thought we could retry non-get requests

aojea commented:
> I'm pretty sure that's not true, we can retry in response to a 429 from the server
> 
> > We also only retry only on GETs ,if I understand the code correctly, that means the request body is never big, we can copy and not have to rewind
> 
> I thought we could retry non-get requests

right, I was wrong, I misread the retry logic, 429 and >= 500 are retryable

MadhavJivrajani commented:
> Adding a unit test to exercise that exact scenario (with a dummy request body that fails Seek()) would be useful.

Working on this.

dims commented:
/retest

MadhavJivrajani commented:
checking if there is an open issue for that unit test. I've noticed it flake a few times in the past couple weeks.

MadhavJivrajani commented:
/retest

tkashem commented:
/assign

MadhavJivrajani commented:
force push includes a fixup that removes the `respCopy` field from the `ResponseCloser` struct

liggitt commented on staging/src/k8s.io/client-go/rest/with_retry.go:
```diff
@@ -372,6 +379,79 @@ func readAndCloseResponseBody(resp *http.Response) {
+		// part of the defer.
+		body := r.resp.Body
+		defer func() {
+			r.err = body.Close()
```
if there's already an error from io.Copy, preserve it:
```
			err := body.Close()
			if r.err == nil {
				r.err = err
			}
```

liggitt commented on staging/src/k8s.io/client-go/rest/with_retry.go:
```diff
@@ -209,14 +211,19 @@ func (r *withRetry) IsNextRetry(ctx context.Context, restReq *Request, httpReq *
 		return ctx.Err()
 	}
 
 	// Ensure the response body is fully read and closed before
 	// we reconnect, so that we reuse the same TCP connection.
```
nit: delete the comment from here, since we moved the comment into the `if` block

liggitt commented:
this change looks right to me. still need the unit test described in https://github.com/kubernetes/kubernetes/pull/109050#issuecomment-1079714763 and would like @smarterclayton's input if possible

MadhavJivrajani commented:
> this change looks right to me. still need the unit test described in [#109050 (comment)](https://github.com/kubernetes/kubernetes/pull/109050#issuecomment-1079714763) and would like @smarterclayton's input if possible

Thanks Jordan. Will working on completing the unit test tomorrow!

liggitt commented:
(also, given our use was apparently always incorrect, I'm wondering if we should expect to eventually pick this to release branches... would like @smarterclayton's input on that as well)

tkashem committed:
client-go: reset request body after response is read and closed

This commit refactors the retry logic to include resetting the
request body. The reset logic will be called iff it is not the
first attempt. This refactor is nescessary mainly because now
as per the retry logic, we always ensure that the request body
is reset *after* the response body is *fully* read and closed
in order to reuse the same TCP connection.

Previously, the reset of the request body and the call to read
and close the response body were not in the right order, which
leads to race conditions.

This commit also adds a test that verifies the order in which
the function calls are made to ensure that we seek only after
the response body is closed.

Co-authored-by: Madhav Jivrajani <madhav.jiv@gmail.com>

tkashem commented:
> Reading the body into a buffer is nescessary because the way Requests are implemented

After looking at the retry logic, looks like the call to reset request body (which is in `IsNextRetry` method) and the call to read and close the response body is not in proper order.
If we fix the order by moving the request body reset logic into the `Before` method then we always ensure that request body is reset after response body is fully read and closed, thus eliminating the need to read the response body into a buffer. 


@MadhavJivrajani  I have drafted the concept into - https://github.com/kubernetes/kubernetes/pull/109088, feel free to refer to it if you want to adopt this approach.

aojea commented:
> (also, given our use was apparently always incorrect, I'm wondering if we should expect to eventually pick this to release branches... would like @smarterclayton's input on that as well)

@liggitt I don't agree with the comment that closing the response body will guarantee the request body stops being read, at least is not how it works today and seems it was this way before too https://github.com/golang/go/issues/51907#issuecomment-1081486340

MadhavJivrajani commented:
@tkashem thank you! Your approach seems to be far cleaner than buffering. Imo it solves the problem at hand. If folks agree, I'm more than happy to go with your approach.

dims commented:
@MadhavJivrajani go for it!

MadhavJivrajani commented:
@tkashem I've incorporated your solution here with a few minor edits (mostly nits), notably, I've changed what the error message says on seek failure since we are now surfacing this.

MadhavJivrajani commented:
/retest

MadhavJivrajani commented:
The above 3 job failures seem to be unrelated, other PRs are facing the same failures. Started a thread on sig-testing: https://kubernetes.slack.com/archives/C09QZ4DQB/p1648560404301529

smarterclayton commented on staging/src/k8s.io/client-go/rest/request_test.go:
```diff
@@ -1130,7 +1139,10 @@ func TestRequestStream(t *testing.T) {
 			Err: true,
 			ErrFn: func(err error) bool {
-				return apierrors.IsInternalError(err)
+				if !strings.Contains(err.Error(), "failed to reset request body for *rest.Request - EOF") {
```
This seems incorrect in general

smarterclayton commented on staging/src/k8s.io/client-go/rest/request_test.go:
```diff
@@ -2474,7 +2485,7 @@ func TestRequestWithRetry(t *testing.T) {
 			body:                         &readSeeker{err: io.EOF},
 			serverReturns:                responseErr{response: nil, err: io.ErrUnexpectedEOF},
-			errExpected:                  io.ErrUnexpectedEOF,
+			errContains:                  "failed to reset request body for *rest.Request - EOF",
```
Is this the right error to return?  Why would the caller want to deal with the wrapper?

smarterclayton replied:
This test is very non-obvious why it fails.  Technically the test is saying "i have an empty body", it's just that Seek now returns EOF, which is abnormal.

As far as the message, this message is probably not accurate.  The programmer has given us a body that is not reusable for retries, which is programmer error OR a transient error.  We probably can't tell the difference easily, so the message needs to reflect that.

I would probably suggest the error message be changed to `failed to reset the request body while retrying a request: EOF`

smarterclayton commented on staging/src/k8s.io/client-go/rest/with_retry.go:
```diff
@@ -245,6 +221,18 @@ func (r *withRetry) Before(ctx context.Context, request *Request) error {
+	if seeker, ok := request.body.(io.Seeker); ok && request.body != nil {
+		if _, err := seeker.Seek(0, io.SeekStart); err != nil {
+			err = fmt.Errorf("failed to reset request body for %T - %w", request, err)
+			r.trackPreviousError(err)
```
Calling trackPreviousError seems suspect here - the error is "we can't read the body" but that's not the previous error anymore?

tkashem replied:
the current error has `EOF` wrapped, so the higher level function should still be able to determine that the error returned (current error) is an EOF error.
We should already have a test for that, if not we should add one.

smarterclayton replied:
As a client, I'm not 100% sure that a failure to seek from a programmer error (passing request a non-startable stream) is an error that should be returned.  However, implementing Seekable is probably as close as we can get to inferring that intent, in which case the "current" error does become "we failed to seek due to a filesystem or other remote error", so I guess this is ok.

smarterclayton commented:
@tkashem Let's take it to the other PR, I'll come back to reviewing this once I understand those changes

smarterclayton commented:
Note that https://github.com/kubernetes/kubernetes/pull/102217/files#r837633535 broke thread safety on methods called on Request from multiple goroutines that have retries, which it should not have.  I'm not sure that's the same issue here, but it is definitely wrong, and I am probably going to ask for a revert of #102217 and #108698 as a consequence, and a redesign that allows those to preserve thread safety.  Will come back to this review after that.

smarterclayton commented on staging/src/k8s.io/client-go/rest/request.go:
```diff
@@ -614,15 +614,14 @@ func (r *Request) Watch(ctx context.Context) (watch.Interface, error) {
 			return nil, r.retry.WrapPreviousError(err)
 		}
 
+		req, err := r.newHTTPRequest(ctx)
```
The addition of `.retry` to the Request object broke threadsafety, and so if we have to remove `retry` this refactor doesn't make sense.  Will come back after discussing this.

smarterclayton replied:
I think thread safety is probably not a release blocker (previous regression), unlike the body close race condition.  Therefore we don't have to block this fix on a refactor of `retry`, the refactor will simply use this logic.

smarterclayton commented:
I might suggest testing whether the race in #108906 occurs if both #102217 and #108698 are reverted locally (just to eliminate that).  @MadhavJivrajani is that something you can try?

smarterclayton commented:
I'm asking for a full revert of #102217 and its successors because it broke retry logic for PUT/POST/DELETE when server errors indicated a retry was possible, and as a consequence of that this PR will have to be updated to deal with that.  Stay tuned

cici37 commented:
/triage accepted

smarterclayton commented:
> I'm asking for a full revert of https://github.com/kubernetes/kubernetes/pull/102217 and its successors because it broke retry logic for PUT/POST/DELETE when server errors indicated a retry was possible from 1.22 onwards, and as a consequence of that this PR will have to be updated to deal with that. Stay tuned

After deeper review the suspected issue was not really an issue - so we're left with the regression to thread safety only (since 1.22) in methods on `Request`.  That *probably* is not a release blocker (it's a regression that could silently break people, but a backport can address it), so I'll review this assuming that code change stays, and then @tkashem will have more time to find a thread safe factoring.

smarterclayton commented on staging/src/k8s.io/client-go/rest/with_retry.go:
```diff
@@ -245,6 +221,18 @@ func (r *withRetry) Before(ctx context.Context, request *Request) error {
+	// the same TCP connection.
+	if seeker, ok := request.body.(io.Seeker); ok && request.body != nil {
+		if _, err := seeker.Seek(0, io.SeekStart); err != nil {
+			err = fmt.Errorf("failed to reset request body for %T - %w", request, err)
```
See https://github.com/kubernetes/kubernetes/pull/109050/files#r838594245 but the fmt error should probably be:

    fmt.Errorf("failed to reset the request body while retrying a request: %v", err)

MadhavJivrajani replied:
@smarterclayton I've addressed this review, thank you! PTAL :)

MadhavJivrajani commented:
/retest

smarterclayton commented on staging/src/k8s.io/client-go/rest/with_retry.go:
```diff
@@ -194,46 +198,18 @@ func (r *withRetry) IsNextRetry(ctx context.Context, restReq *Request, httpReq *
-}
-
 func (r *withRetry) Before(ctx context.Context, request *Request) error {
+	// If the request context is already canceled there
```
I had a question about After for @tkashem here https://github.com/kubernetes/kubernetes/pull/108347/files#r838810333 - it's not clear to me what the value of After is at this point (may be something his refactoring to fix thread safety in Request will have to deal with)

MadhavJivrajani replied:
+1, I've been wondering the same. 
Would it maybe make sense to put reading and closing response body as part of `After()`? 
Seems intuitive in the flow of `Before` -> attempt -> `After`
and the caller doesn't have to handle it explicitly because the implementation would do it for them (and considering forgetting to implement it could lead to races again, maybe we could consider making reading and closing part of the implementation itself)

tkashem replied:
> Would it maybe make sense to put reading and closing response body as part of After()?

it will be a little tricky to achieve that given that we have transform function, but definitely something we can look into during the follow-up refactor.

for now, we need `After` to update the backoff which needs the reply tuple (response, err) from the server.

MadhavJivrajani replied:
oh 100%, definitely part of a follow up!

smarterclayton commented on staging/src/k8s.io/client-go/rest/request_test.go:
```diff
@@ -938,7 +938,10 @@ func TestRequestWatch(t *testing.T) {
 			Err: true,
 			ErrFn: func(err error) bool {
-				return apierrors.IsInternalError(err)
+				if !strings.Contains(err.Error(), "failed to reset the request body while retrying a request: EOF") {
```
I think we *may* need to assert that this is `!apierrors.IsInternalError(err)`, or assert this is not a server error at all (possibly by checking what the type is after unwrapping).

MadhavJivrajani commented:
/retest

smarterclayton commented:
Do we have enough testing to be confident the race is now closed with the current changes?  If so, I'll tag this now, otherwise I'll wait for that.

tkashem commented:
> Do we have enough testing to be confident the race is now closed with the current changes? If so, I'll tag this now, otherwise I'll wait for that.

if the race happens because we seek the request body before we close the response body, then we can add the following test:
```
// TODO: see if we can consolidate the other trackers into one.
type requestBodyTracker struct {
	io.ReadSeeker
	f func(string)
}

func (t *requestBodyTracker) Read(p []byte) (int, error) {
	t.f("Request.Body.Read")
	return t.ReadSeeker.Read(p)
}

func (t *requestBodyTracker) Seek(offset int64, whence int) (int64, error) {
	t.f("Request.Body.Seek")
	return t.ReadSeeker.Seek(offset, whence)
}

type responseBodyTracker struct {
	io.ReadCloser
	f func(string)
}

func (t *responseBodyTracker) Read(p []byte) (int, error) {
	t.f("Response.Body.Read")
	return t.ReadCloser.Read(p)
}

func (t *responseBodyTracker) Close() error {
	t.f("Response.Body.Close")
	return t.ReadCloser.Close()
}

type recorder struct {
	order []string
}

func (r *recorder) record(call string) {
	r.order = append(r.order, call)
}

func TestRequestBodyResetOrder(t *testing.T) {
	recorder := &recorder{}
	respBodyTracker := &responseBodyTracker{
		ReadCloser: nil, // the server will fill it
		f:          recorder.record,
	}

	var attempts int
	client := clientForFunc(func(req *http.Request) (*http.Response, error) {
		defer func() {
			attempts++
		}()

		// read the request body.
		ioutil.ReadAll(req.Body)

		// first attempt, we send a retry-after
		if attempts == 0 {
			resp := retryAfterResponse()
			respBodyTracker.ReadCloser = ioutil.NopCloser(bytes.NewReader([]byte{}))
			resp.Body = respBodyTracker
			return resp, nil
		}

		return &http.Response{StatusCode: http.StatusOK}, nil
	})

	reqBodyTracker := &requestBodyTracker{
		ReadSeeker: bytes.NewReader([]byte{}), // empty body ensures one Read operation at most.
		f:          recorder.record,
	}
	req := &Request{
		verb: "POST",
		body: reqBodyTracker,
		c: &RESTClient{
			content: defaultContentConfig(),
			Client:  client,
		},
		backoff: &noSleepBackOff{},
		retry:   &withRetry{maxRetries: 1},
	}

	req.Do(context.Background())

	expected := []string{
		// 1st attempt: the server handler reads the request body
		"Request.Body.Read",
		// the server sends a retry-after, client reads the
		// response body, and closes it
		"Response.Body.Read",
		"Response.Body.Close",
		// client retry logic seeks to the beginning of the request body
		"Request.Body.Seek",
		// 2nd attempt: the server reads the request body
		"Request.Body.Read",
	}
	if !reflect.DeepEqual(expected, recorder.order) {
		t.Errorf("Expected invocation request and response body operations for retry do not match: %s", cmp.Diff(expected, recorder.order))
	}
}
```

This test verifies that request body is reset with a `Seek` only after the response body is read and closed. If I run the test against master it fails with
```
$ go test -race -v -count=1 k8s.io/client-go/rest -run=TestRequestBodyResetOrder
=== RUN   TestRequestBodyResetOrder
    request_test.go:3631: Expected invocation request and response body operations for retry do not match:   []string{
          	"Request.Body.Read",
        + 	"Request.Body.Seek",
          	"Response.Body.Read",
          	"Response.Body.Close",
        - 	"Request.Body.Seek",
          	"Request.Body.Read",
          }
--- FAIL: TestRequestBodyResetOrder (0.00s)
FAIL
FAIL	k8s.io/client-go/rest	0.101s
FAIL
```
with this PR, it passes as expected.

@MadhavJivrajani you can add this test to the PR if you want.

tkashem commented:
Thanks @dims 
I also stressed the test, looks fine so far:
```
$ stress ./rest.test -test.run=TestRequestBodyResetOrder
5s: 575 runs so far, 0 failures
10s: 1191 runs so far, 0 failures
15s: 1821 runs so far, 0 failures
20s: 2398 runs so far, 0 failures
...
3m15s: 23940 runs so far, 0 failures
3m20s: 24537 runs so far, 0 failures
```

aojea commented:
closing the response body doesn't guarantee that the request body has been closed ... but it seems that it was always that way and is a very edge case, so I don't worry about that, just be sure we close the response body before Seek

https://github.com/golang/go/issues/51907#issuecomment-1079707826

smarterclayton commented:
would be good to see that test added, this is lgtm from my perspective, but I’ll let one other person do the tag when the test is added (@tkashem probably)

/approve

MadhavJivrajani commented:
I've added the test, thank you @tkashem and @smarterclayton!

MadhavJivrajani commented:
/retest

tkashem commented:
seems unrelated to this PR

/test pull-kubernetes-node-e2e-containerd

MadhavJivrajani commented:
/retest
Should be fine now
https://kubernetes.slack.com/archives/CCK68P2Q2/p1649083731159209?thread_ts=1649083731.159209&cid=CCK68P2Q2

palnabarun commented:
/retest

tkashem commented:
/test pull-kubernetes-node-e2e-containerd

MadhavJivrajani commented:
/retest
Timeout

MadhavJivrajani commented:
/test pull-kubernetes-e2e-gce-100-performance

palnabarun commented:
/test pull-kubernetes-e2e-gce-100-performance
